### PR TITLE
REF: do extract_array earlier in series arith/comparison ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
With this, the middle third of _arith_method_SERIES and _comp_method_SERIES are array-specific and can be refactored out (separate step) to become a) block-wise implementation for DataFrame and b) PandasArray implementation.

This also simplifies the NullFrequencyError handling nicely.

### CODE REVIEW

File: pandas/core/ops/__init__.py [hunk 1]
```diff
@@ -520,13 +524,29 @@ def column_op(a, b):
     return result
 
 
-def dispatch_to_extension_op(op, left, right):
+def dispatch_to_extension_op(op, left, right, keep_null_freq: bool = False):
```
Comment (jbrockmendel) [inline pandas/core/ops/__init__.py]: @simonjayhawkins do we have a way of typing `left` and `right` as "not a Series or Index"?

**Comment by jbrockmendel** [comment]
@jreback gentle ping; this turns out to be a blocker(ish) for fixing #28080.

File: pandas/core/ops/__init__.py [hunk 2]
```diff
@@ -157,12 +160,10 @@ def maybe_upcast_for_op(obj, shape: Tuple[int, ...]):
...
             #  which would incorrectly be treated as a datetime-NaT, so
-            #  we broadcast and wrap in a Series
+            #  we broadcast and wrap in a TimedeltaArray
+            obj = obj.astype("timedelta64[ns]")
             right = np.broadcast_to(obj, shape)
-
-            # Note: we use Series instead of TimedeltaIndex to avoid having
-            #  to worry about catching NullFrequencyError.
-            return pd.Series(right)
+            return TimedeltaArray(right)
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: i don’t recall but do we copy by default?

File: pandas/core/ops/__init__.py (hunk 1 above)
Comment (jreback) [reply pandas/core/ops/__init__.py]: can u type left here (EA / np.ndarray)

File: pandas/core/ops/__init__.py [hunk 3]
```diff
@@ -615,25 +630,29 @@ def wrapper(left, right):
...
-            # We should only get here with td64 right with non-scalar values
-            #  for right upcast by maybe_upcast_for_op
-            assert not isinstance(right, (np.timedelta64, np.ndarray))
-            result = op(left._values, right)
+        rvalues = maybe_upcast_for_op(rvalues, lvalues.shape)
 
-        else:
-            lvalues = extract_array(left, extract_numpy=True)
-            rvalues = extract_array(right, extract_numpy=True)
+        if should_extension_dispatch(left, rvalues):
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: what is this check for? isn’t by definition we always should dispatch if we have an EA left?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops/__init__.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/ops/__init__.py]: > @simonjayhawkins do we have a way of typing `left` and `right` as "not a Series or Index"?

i'm not aware of being able to exclude types.

if a particular type raises, (in this case TYpeError?) then maybe could use overloads with a return type of NoReturn https://mypy.readthedocs.io/en/latest/more_types.html#the-noreturn-type (New in version 3.5.4)

could maybe use the following pattern to allow checking with older Python...

```
if TYPE_CHECKING:
    from typing import NoReturn
else:
    NoReturn = None
```

File: pandas/core/ops/__init__.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: no we dont

File: pandas/core/ops/__init__.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/ops/__init__.py]: That's what this is: checking if `left` is an EA, or if any of the other conditions for dispatching hold

**Comment by jbrockmendel** [comment]
Types added

**Review by TomAugspurger — APPROVED** [review]
Since `keep_null_freq` is just a temporary thing until the deprecation can be enforced, I'm +1 on this.

I'd recommend giving @jreback about 48 hours to look, and then merge if he hasn't had a chance by them.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel 

yeah the ``keep_null_freq`` stuff is a bit hard to follow, though you have documented it, so I think ok
